### PR TITLE
Recommendations

### PR DESCRIPTION
**Pull request by juliwang143**
Description:

- added recent searches routes to backend for adding recent searches, getting recent searches, and clearing recent searches when a user logs out
- ensured that at most the top 10 most recent searches are stored
- revised similarity score between users now consists of weighing the number of shared artist/genre preferences, audio analysis of posted songs, and similarities in recent searches

Milestone: Technical Problem - Recommendations

Test Plan:

Screen Recording 2022-07-27 at 10.52.45 AM.mov

### CODE REVIEW

File: capstone-ui/src/components/Search/Search.jsx
Comment (nltran12) [inline capstone-ui/src/components/Search/Search.jsx]: handle error?

File: capstone-api/models/User.js
Comment (nltran12) [inline capstone-api/models/User.js]: I think you can put this in if (user.get("recentSearches").length >= 10)  block since it only should run if you need to delete an old search. Also you don't need to instantiate oldestSearch outside of the if section.

File: capstone-ui/src/recommendationUtils.js
Comment (nltran12) [inline capstone-ui/src/recommendationUtils.js:12]: could you add a comment about what this function does?

File: capstone-ui/src/recommendationUtils.js
Comment (nltran12) [inline capstone-ui/src/recommendationUtils.js:32]: Also add comment about what this function does

**Review by nltran12 — CHANGES_REQUESTED** [review]
A few change requests, the additional code looks awesome though! Let's also wait for Mulan to review

File: capstone-api/routes/User.js
Comment (nltran12) [inline capstone-api/routes/User.js]: nit: instead of result, let's name variables more explicitly.  instead of result, you could name this isAddRecentSearchSuccessful. It's okay if variables are a bit long-- as long as it's easier for others to understand

File: capstone-ui/src/components/Search/Search.jsx
Comment (nltran12) [inline capstone-ui/src/components/Search/Search.jsx]: suggestion on rewriting this as:
const oldSearchResults = searchInput == "" ? songResults : [];
const trueSearchValue = searchInput == ""  ? searchInputValue ? "";

File: capstone-ui/src/components/Search/Search.jsx
Comment (nltran12) [inline capstone-ui/src/components/Search/Search.jsx:76]: This can be 1 line since it doesn't seem like you actually use newResults
const addedResults = oldSearchResults.concat(data.tracks.items);

File: capstone-ui/src/components/Settings/Settings.jsx
Comment (nltran12) [inline capstone-ui/src/components/Settings/Settings.jsx]: recommendations != null &&

File: capstone-ui/src/components/SongCard/SongCard.jsx
Comment (nltran12) [inline capstone-ui/src/components/SongCard/SongCard.jsx]: delete

File: capstone-ui/src/components/SongHeaderView/SongHeaderView.jsx
Comment (nltran12) [inline capstone-ui/src/components/SongHeaderView/SongHeaderView.jsx]:
```suggestion
        </iframe
          className="songheader-playing-picture"
          src={`https://open.spotify.com/embed/track/${song.id}`}
          width="100%"
          height="380"
          frameBorder="0"
        >
```

File: capstone-ui/src/components/SongHeader/SongHeader.jsx
Comment (nltran12) [inline capstone-ui/src/components/SongHeader/SongHeader.jsx]:
```suggestion
          </iframe
            className="songheader-playing-picture"
            src={`https://open.spotify.com/embed/track/${song.id}`}
            width="100%"
            height="380"
            frameBorder="0"
          >
```

**Review by nltran12 — CHANGES_REQUESTED** [review]
Some more style suggestions, also  let's wait for mulan to review

File: capstone-api/models/User.js
Comment (mulanz) [inline capstone-api/models/User.js]: nit: What is "result" for? This could be named more clearly!

File: capstone-api/models/User.js
Comment (mulanz) [inline capstone-api/models/User.js]: For comments like these, you don't have to describe how you are doing it (especially since this is relatively simple and people can probably figure out that you are removing the first element to get the first element to delete). Instead, you can talk about your thought process for doing this: e.g. "We are keeping the recent searches under 10 items."

File: capstone-api/models/User.js
Comment (mulanz) [inline capstone-api/models/User.js]: I think this specifically doesn't need a comment!

File: capstone-ui/src/components/Error/Error.css
Comment (mulanz) [inline capstone-ui/src/components/Error/Error.css]: this is great!

File: capstone-ui/src/components/ProfileHeader/ProfileHeader.jsx
Comment (mulanz) [inline capstone-ui/src/components/ProfileHeader/ProfileHeader.jsx]: Not sure if you've already refactored or not since you took out "TODO"; Either way, this comment is not very clear what's going on, so I think you can either take it out or make it a bit more clear. "We are checking for username and profile view in order to...."

File: capstone-ui/src/components/ProfileHeader/ProfileHeader.jsx
Comment (mulanz) [inline capstone-ui/src/components/ProfileHeader/ProfileHeader.jsx]: I don't think this comment is needed actually!

File: capstone-ui/src/components/Search/Search.jsx
Comment (mulanz) [reply capstone-ui/src/components/Search/Search.jsx:76]: also, would make sense to name it updatedResults; addedResults doesn't sound like you concatted it, it sounds like it's only the new ones you added

File: capstone-ui/src/components/SongCard/SongCard.css
Comment (mulanz) [inline capstone-ui/src/components/SongCard/SongCard.css]: did you mean to take this out?

File: capstone-ui/src/recommendationUtils.js
Comment (mulanz) [inline capstone-ui/src/recommendationUtils.js:31]: great!

File: capstone-ui/src/recommendationUtils.js
Comment (mulanz) [inline capstone-ui/src/recommendationUtils.js]: amazing

File: capstone-ui/src/spotify.js
Comment (mulanz) [inline capstone-ui/src/spotify.js]: no need to comment this

**Review by mulanz — COMMENTED** [review]
this looks so great! I'm so impressed by the complexity of the recommender, nice job with all the organization. just some nits

File: capstone-ui/src/components/Search/Search.jsx
Comment (juliwang143) [reply capstone-ui/src/components/Search/Search.jsx]: I fixed this in my Search PR just now so when I merge I will make sure to choose that change

File: capstone-ui/src/components/Search/Search.jsx
Comment (juliwang143) [reply capstone-ui/src/components/Search/Search.jsx:76]: setSongResults(oldResults.concat(data.tracks.items));
I updated this in my Search PR so that the line now is the above ^

File: capstone-ui/src/components/Settings/Settings.jsx
Comment (juliwang143) [reply capstone-ui/src/components/Settings/Settings.jsx]: Fixed in the play feature PR will chose that change when merging

File: capstone-ui/src/components/SongCard/SongCard.jsx
Comment (juliwang143) [reply capstone-ui/src/components/SongCard/SongCard.jsx]: deleted in other PR will chose that change when merging

File: capstone-ui/src/components/SongHeaderView/SongHeaderView.jsx
Comment (juliwang143) [reply capstone-ui/src/components/SongHeaderView/SongHeaderView.jsx]: I will go through all my components after merging them all to format them like this

File: capstone-ui/src/components/ProfileHeader/ProfileHeader.jsx
Comment (juliwang143) [reply capstone-ui/src/components/ProfileHeader/ProfileHeader.jsx]: removed in other PR

File: capstone-ui/src/components/SongCard/SongCard.css
Comment (juliwang143) [reply capstone-ui/src/components/SongCard/SongCard.css]: removed in other PR so I will keep that change when merging

File: capstone-ui/src/components/Search/Search.jsx
Comment (nltran12) [reply capstone-ui/src/components/Search/Search.jsx:76]: nice refactoring

**Review by nltran12 — APPROVED** [review]
lgtm

**Pull request merged**
